lib/Parse/ParseGeneric.cpp

```diff
@@ -261,22 +262,46 @@ ParserStatus Parser::parseGenericWhereClause(
     if (Tok.is(tok::colon)) {
       // A conformance-requirement.
       SourceLoc ColonLoc = consumeToken();
-
-      // Parse the protocol or composition.
-      ParserResult<TypeRepr> Protocol = parseTypeForInheritance(
-          diag::expected_identifier_for_type,
-          diag::expected_ident_type_in_inheritance);
-      
-      if (Protocol.isNull()) {
-        Status.setIsParseError();
-        if (Protocol.hasCodeCompletion())
-          Status.setHasCodeCompletion();
-        break;
+      Identifier TypeId;
+
+      if (Tok.is(tok::identifier) &&
+          getLayoutConstraintInfo(TypeId = Context.getIdentifier(Tok.getText()),
```

Comment:
This would be more readable if it was split into separate expressions in my opinion